compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -317,13 +317,25 @@ fn layout_of_uncached<'tcx>(
             if count.has_aliases() {
                 count = tcx.normalize_erasing_regions(cx.typing_env, count);
                 if count.has_aliases() {
-                    return Err(error(cx, LayoutError::Unknown(ty)));
+                    return Err(error(cx, LayoutError::TooGeneric(ty)));
                 }
             }
 
-            let count = count
-                .try_to_target_usize(tcx)
-                .ok_or_else(|| error(cx, LayoutError::Unknown(ty)))?;
+            let count = match count.kind() {
```

Comment:
There are some other constants being evaluated above for the unstable pattern types:

```
.try_to_bits(tcx, cx.typing_env)
.ok_or_else(|| error(cx, LayoutError::Unknown(ty)))?;
```

These should also be handled in a similar manner to the array lengths here.

Reply:
We should check the `try_to_bits` only if the `ConstKind` is actually a `Value` right?. More precisely, the valtree associated.

Reply:
Yes we should only call `try_to_bits` if `start` / `end` are `ConstKind::Value`, if that's what you're asking.

Basically everything else there should follow the same logic as here, except replace `try_to_target_usize` with `try_to_bits`. (So it may make sense to extract the common logic into a function.)